runtime/src/main/scala/akka/grpc/internal/PlayRouterUsingActions.scala

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import akka.annotation.InternalApi
+import akka.grpc.Grpc
+import akka.http.scaladsl.model.HttpHeader.ParsingResult
+import akka.http.scaladsl.model._
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import play.api.http.HttpChunk
+import play.api.http.HttpChunk.{ Chunk, LastChunk }
+import play.api.http.HttpEntity.Chunked
+import play.api.libs.streams.Accumulator
+import play.api.mvc._
+import play.api.routing.Router
+import play.api.routing.Router.Routes
+
+import scala.collection.immutable
+import scala.concurrent.{ ExecutionContext, Future }
+
+/**
+ * Boiler plate needed for the generated Play routers allowing for adding a service implementation in a Play app,
+ * inherited by the generated abstract service router (both Java and Scala) which is then implemented by the user.
+ *
+ * INTERNAL API
+ */
+@InternalApi abstract class PlayRouterUsingActions(mat: Materializer, serviceName: String, parsers: PlayBodyParsers, actionBuilder: DefaultActionBuilder)
+  extends play.api.routing.Router {
+
+  private val prefix = s"/$serviceName"
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  protected def createHandler(
+    serviceName: String,
+    mat: Materializer): RequestHeader => EssentialAction
+
+  private val handler = createHandler(serviceName, mat)
+
+  // Scala API
+  final override def routes: Routes = {
+    case rh if rh.path.startsWith(prefix) ⇒ handler(rh)
+  }
+
+  final override def documentation: Seq[(String, String, String)] = Seq.empty
+
+  /**
+   * Registering a gRPC service under a custom prefix is not widely supported and strongly discouraged by the specification
+   * so therefore not supported.
+   */
+  final override def withPrefix(prefix: String): Router =
+    if (prefix == "/") this
+    else
+      throw new UnsupportedOperationException(
+        "Prefixing gRPC services is not widely supported by clients, " +
+          s"strongly discouraged by the specification and therefore not supported. " +
+          s"Attempted to prefix with [$prefix], yet already default prefix known to be [${this.prefix}]. " +
+          s"When binding gRPC routers the path in `routes` MUST BE `/`.")
+
+  def createStreamingAction(handler: HttpRequest => Future[HttpResponse])(implicit ec: ExecutionContext): EssentialAction =
+    actionBuilder.async(streamBodyParser) { req =>
+      handler(playToAkkaRequestStream(req)).map(akkaToPlayResp)
```

Comment:
As discussed in https://github.com/akka/akka-grpc/issues/420#issuecomment-426630873, this 'additional roundtrip' is of course more or less unavoidable if you want to take into account the changes to the Play request, but does come at some performance cost.

How much performance cost, and whether that cost is acceptable, of course highly depends on the application and should be measured/considered by the user.

I think we should add some documentation to this option, and explain clearly that the trade-off here: this option allows you to use your existing Play action composition approach, at the expense of some performance (which is why it is not the default). Later, eventually, we could also document alternate ways to achieve cross-cutting concerns there.

Reply:
Currently there's no mention of the Play integration in the docs. Where do you suggest documenting this?

Reply:
Ah, the Play gRPC documentation has already moved to https://github.com/playframework/play-grpc - I guess we will have to move the documentation for this feature to a separate PR over there.

Reply:
Ok, I'll work on that before long, but you can go ahead and merge this right?